types/wlr_surface_layers.c

```diff
@@ -0,0 +1,287 @@
+#include <stdlib.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/types/wlr_surface.h>
+#include <wlr/types/wlr_surface_layers.h>
+#include <wlr/util/log.h>
+#include "surface-layers-protocol.h"
+
+static const char *surface_layers_role = "layer_surface";
+
+#define WLR_LAYER_SURFACE_INPUT_DEVICE_COUNT 3
+
+static void layer_surface_set_interactivity(struct wl_client *client,
+		struct wl_resource *resource, uint32_t input_types,
+		uint32_t exclusive_types) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+
+	if (input_types >> WLR_LAYER_SURFACE_INPUT_DEVICE_COUNT) {
+		wl_resource_post_error(resource,
+			LAYER_SURFACE_ERROR_INVALID_INPUT_DEVICE,
+			"Unknown input device in input_types");
+		return;
+	}
+	if (exclusive_types >> WLR_LAYER_SURFACE_INPUT_DEVICE_COUNT) {
+		wl_resource_post_error(resource,
+			LAYER_SURFACE_ERROR_INVALID_INPUT_DEVICE,
+			"Unknown input device in exclusive_types");
+		return;
+	}
+
+	surface->pending->input_types = input_types;
+	surface->pending->exclusive_types = exclusive_types;
+	surface->pending->invalid |= WLR_LAYER_SURFACE_INVALID_INTERACTIVITY;
+}
+
+static void layer_surface_set_anchor(struct wl_client *client,
+		struct wl_resource *resource, uint32_t anchor) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+
+	if (anchor >> 4) {
+		wl_resource_post_error(resource, LAYER_SURFACE_ERROR_INVALID_ANCHOR,
+			"Unknown anchor");
+		return;
+	}
+
+	surface->pending->anchor = anchor;
+	surface->pending->invalid |= WLR_LAYER_SURFACE_INVALID_ANCHOR;
+}
+
+static void layer_surface_set_exclusive_zone(struct wl_client *client,
+		struct wl_resource *resource, uint32_t zone) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->pending->exclusive_zone = zone;
+	surface->pending->invalid |= WLR_LAYER_SURFACE_INVALID_EXCLUSIVE_ZONE;
+}
+
+static void layer_surface_set_margin(struct wl_client *client,
+		struct wl_resource *resource, int32_t horizontal, int32_t vertical) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->pending->margin_horizontal = horizontal;
+	surface->pending->margin_vertical = vertical;
+	surface->pending->invalid |= WLR_LAYER_SURFACE_INVALID_MARGIN;
+}
+
+void wlr_layer_surface_move_state(struct wlr_layer_surface_state *next,
+		struct wlr_layer_surface_state *state) {
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_INTERACTIVITY) {
+		state->input_types = next->input_types;
+		state->exclusive_types = next->exclusive_types;
+	}
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_ANCHOR) {
+		state->anchor = next->anchor;
+	}
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_EXCLUSIVE_ZONE) {
+		state->exclusive_zone = next->exclusive_zone;
+	}
+	if (next->invalid & WLR_LAYER_SURFACE_INVALID_MARGIN) {
+		state->margin_horizontal = next->margin_horizontal;
+		state->margin_vertical = next->margin_vertical;
+	}
+
+	state->invalid |= next->invalid;
```

Comment:
probably better to call this `changed`.

Reply:
I think we use invalid elsewhere

Reply:
It's definitely wrong here, and probably wrong elsewhere too.

Reply:
How do you figure? Invalidate is a common term in systems for when some kind of state is changed and has to be flushed somewhere.

Reply:
I guess you're right. Like "cache invalidation". These are the hard things in computer science I guess.

Reply:
It's only hard when you invent new terms, using established ones is easy :wink: